examples/kalman_filter.ipynb

```diff
@@ -0,0 +1,490 @@
+    "        solver=solver,\n",
+    "        stepsize_controller=stepsize_controller,\n",
+    "        t0=ts[0],\n",
+    "        t1=ts[-1],\n",
+    "        y0=y0,\n",
+    "        dt0=dt0,\n",
+    "        saveat=dfx.SaveAt(ts=ts),\n",
+    "    ).ys\n",
+    "\n",
+    "\n",
+    "def simulate_lti_system(\n",
+    "    sys: LTISystem,\n",
+    "    y0: jnp.ndarray,\n",
+    "    ts: jnp.ndarray,\n",
+    "    us: Optional[jnp.ndarray] = None,\n",
+    "    std_measurement_noise: float = 0.0,\n",
+    "    key=jr.PRNGKey(\n",
+    "        1,\n",
+    "    ),\n",
+    "):\n",
+    "    u_t = _interpolate_us(ts, us, sys.B)\n",
+    "\n",
+    "    def rhs(t, y, args):\n",
+    "        return sys.A @ y + sys.B @ u_t.evaluate(t)\n",
+    "\n",
+    "    xs = _diffeqsolve(rhs, ts, y0)\n",
+    "    # noisy measurements\n",
+    "    ys = xs @ sys.C.transpose()\n",
+    "    ys = ys + jr.normal(key, shape=ys.shape) * std_measurement_noise\n",
+    "    return xs, ys"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Here we define the Kalman-Filter.\n",
+    "\n",
+    "Note how we use `equniox` to combine the Kalman-Filter logic in `__call__` and the Kalman-Filter parameters `Q` , `R` in one object."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class KalmanFilter(eqx.Module):\n",
+    "    \"\"\"Continuous-time Kalman Filter\n",
+    "\n",
+    "    Ref:\n",
+    "        [1] Optimal and robust estimation. 2nd edition. Page 154.\n",
+    "        https://lewisgroup.uta.edu/ee5322/lectures/CTKalmanFilterNew.pdf\n",
+    "    \"\"\"\n",
+    "\n",
+    "    sys: LTISystem\n",
+    "    x0: jnp.ndarray\n",
+    "    P0: jnp.ndarray\n",
+    "    Q: jnp.ndarray\n",
+    "    R: jnp.ndarray\n",
+    "\n",
+    "    def __call__(self, ts, ys, us: Optional[jnp.ndarray] = None):\n",
+    "\n",
+    "        A, B, C = self.sys\n",
+    "\n",
+    "        y_t = dfx.LinearInterpolation(ts=ts, ys=ys)\n",
+    "        u_t = _interpolate_us(ts, us, B)\n",
+    "\n",
+    "        n = len(self.x0)\n",
+    "        y0 = jnp.concatenate((self.x0, self.P0.reshape(n**2)))\n",
+    "\n",
+    "        def rhs(t, y, args):\n",
+    "            x, P = y[:n], y[n:].reshape((n, n))\n",
+    "\n",
+    "            # eq 3.22 of Ref [1]\n",
+    "            K = P @ C.transpose() @ jnp.linalg.inv(self.R)\n",
+    "\n",
+    "            # eq 3.21 of Ref [1]\n",
+    "            dPdt = (\n",
+    "                A @ P\n",
+    "                + P @ A.transpose()\n",
+    "                + self.Q\n",
+    "                - P @ C.transpose() @ jnp.linalg.inv(self.R) @ C @ P\n",
+    "            )\n",
+    "\n",
+    "            # eq 3.23 of Ref [1]\n",
+    "            dxdt = A @ x + B @ u_t.evaluate(t) + K @ (y_t.evaluate(t) - C @ x)\n",
+    "\n",
+    "            return jnp.concatenate((dxdt, dPdt.reshape(n**2)))\n",
```

Comment:
Diffrax supports using PyTree-valued state! We don't need to concatenate/reshape/slice here.

`y0 = (self.x0, self.P0)` etc.

Reply:
Heh did not know that. That's nice